### PR TITLE
Problem: fasten unlucky tx patch

### PR DESCRIPTION
**Pull request by mmsqe**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 1]
```diff
@@ -361,7 +382,96 @@ func (db *tmDB) replayTx(appCreator func() *app.App, height int64, txIndex int,
...
+		return nil, err
+	}
+	blockRes.DeliverTxs[res.Index] = &res.Result
+	if err := db.stateStore.SaveABCIResponses(res.Height, &blockRes); err != nil {
+		return nil, err
+	}
+	return &res, nil
+}
+
+func (db *tmDB) patchDB(blockResult *tmstate.ABCIResponses, result *abci.TxResult, height int64, exportOnly bool) (resultErr error) {
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:428]: It's simpler to merge `blockResult` and `result` into one struct, and dump to stdout directly, one line per block.

Comment (mmsqe) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:428]: can't merge 2 proto files, wrong when encode

Comment (yihuang) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:462]: https://github.com/cosmos/cosmos-sdk/blob/main/snapshots/stream.go#L40
try sth like this, `protoio.NewDelimitedWriter` and `protoio.NewDelimitedReader`, stream of `proto.Message`:
```
tmstate.ABCIResponses
abci.TxResult
tmstate.ABCIResponses
abci.TxResult
...
```

Comment (mmsqe) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:462]: I can try, not sure if it's faster

Comment (yihuang) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:462]: it can be managed in a single stream, no need to write to some fixed filenames.

Comment (mmsqe) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:462]: stream seems work, we persist to single file?

**Review by yihuang — CHANGES_REQUESTED** [review]
It should be simply a `protoio.NewDelimitedWriter(io.Stdout)`, and write all the messages into the same stream.
- no need the compression/chunking stuff
- no need separate files per block
- don't even need a fixed filename, just dump to stdout, let user pick destination file with pipe

**Comment by mmsqe** [comment]
> It should be simply a `protoio.NewDelimitedWriter(io.Stdout)`, and write all the messages into the same stream.

I was thinking the same, but don't know how to read back for each block

**Comment by yihuang** [comment]
> > It should be simply a `protoio.NewDelimitedWriter(io.Stdout)`, and write all the messages into the same stream.
> 
> I was thinking the same, but don't know how to read back for each block

I think it's sth like this:

Writer side:
```
for {
  writer.WriteMsg(&abciResponses)
  writer.WriteMsg(&txResult)
}
```

```
var abciResponses ABCIResponses
var txResult TxResult
for {
  protoReader.ReadMsg(&abciResponses)
  protoReader.ReadMsg(&txResult)
}
```

**Comment by mmsqe** [comment]
> Writer side:

when handle single yes, but not thread safe for multi

**Comment by yihuang** [comment]
> > Writer side:
> 
> when handle single yes, but not thread safe for multi

Use a single coroutine to write, pass the pair through channel.

**Comment by yihuang** [comment]
Can you support stdin/stdout, they are friendly to scripts, like:
- `--export-to-file -` means dump to stdout
- `--patch-from-file -` means read from stdin

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 2]
```diff
@@ -101,14 +119,34 @@ func FixUnluckyTxCmd() *cobra.Command {
 					func(baseApp *baseapp.BaseApp) { baseApp.SetCMS(cms) },
 				)
 			}
-
 			// replay and patch a single block
-			processBlock := func(height int64) error {
+			action := "patched"
+			var f *os.File
+			if exportToFile != "" {
+				f, err = os.OpenFile(exportToFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:126]: `O_APPEND` by default may not be intuitive, I would expect it'll override the output file.
If we support stdout, user can append by:
`cronosd --export-to-file - >> /tmp/output`

Comment (mmsqe) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:126]: is it ok `f, err = os.Create(exportToFile)` since higher permission 0666

Comment (yihuang) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:126]: I think `os.Create` is better, as long as we support stdout, the user can do append in the shell anyway.

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@
 - [#470](https://github.com/crypto-org-chain/cronos/pull/470) Run the tasks in `fix-unlucky-tx` command in parallel.
 - [#472](https://github.com/crypto-org-chain/cronos/pull/472) `fix-unlucky-tx` command supports `null` tendermint tx indexer.
 - [#524](https://github.com/crypto-org-chain/cronos/pull/524) Fix tendermint duplicated tx issue.
+- [#571](https://github.com/crypto-org-chain/cronos/pull/571) Support stdout and stdin patch result.
```
Comment (yihuang) [inline CHANGELOG.md:10]:
```suggestion
- [#571](https://github.com/crypto-org-chain/cronos/pull/571) Support export/import patched result, much faster to import the results than replay the tx.
```

**Review by adu-crypto — APPROVED** [review]
lgtm

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 4]
```diff
@@ -96,31 +115,67 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+			std := "-"
 			// replay and patch a single block
-			processBlock := func(height int64) error {
+			if patchFromFile != "" {
+				if patchFromFile == std {
+					err := tmDB.PatchFromImport(clientCtx.TxConfig, os.Stdin)
+					if err != nil {
+						return err
+					}
+					return nil
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:131]:
```suggestion
					return tmDB.PatchFromImport(clientCtx.TxConfig, os.Stdin)
```

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 5]
```diff
@@ -96,31 +115,67 @@ func FixUnluckyTxCmd() *cobra.Command {
...
 				cms.SetLazyLoading(true)
 				return app.New(
-					log.NewNopLogger(), appDB, nil, false, nil,
+					tmlog.NewNopLogger(), appDB, nil, false, nil,
 					ctx.Config.RootDir, 0, encCfg, ctx.Viper,
 					func(baseApp *baseapp.BaseApp) { baseApp.SetCMS(cms) },
 				)
 			}
-
+			std := "-"
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:123]: make it a constant?

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 6]
```diff
@@ -96,31 +115,67 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+					return err
+				}
+				defer func() {
+					returnErr = fi.Close()
+				}()
+				err = tmDB.PatchFromImport(clientCtx.TxConfig, fi)
+				if err != nil {
+					return err
+				}
+				return nil
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:144]:
```suggestion
				return tmDB.PatchFromImport(clientCtx.TxConfig, fi)
```

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 7]
```diff
@@ -96,31 +115,67 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+			}
+			action := "patched"
+			var f *os.File
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:147]: Use `io.Writer` you can assign both file object and stdout to it.

**Review by yihuang — APPROVED** [review]
Some coding style suggestions, and better to test on a production node, otherwise LGTM.

**Comment by mmsqe** [comment]
> Some coding style suggestions, and better to test on a production node, otherwise LGTM.

not sure which node I try on

**Review by yihuang — APPROVED** [review]
LGTM

**Comment by yihuang** [comment]
Can you sign the commits?

**Comment by yihuang** [comment]
Can you fix the conflicts?

**Pull request merged**